### DISCUSSION TITLE
UnicodeDecodeError on conda init

### DISCUSSION THREAD

**Discussion by WulfHans** [idaholab/moose#27278] [Q&A Getting Started] [answered]
When trying a fresh setup, conda  init fails with:
```
rschwa@leander:/home/stud/rschwa$ conda init --all

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/cli/main.py", line 83, in main_subshell
        exit_code = do_call(args, parser)
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/cli/conda_argparse.py", line 196, in do_call
        result = getattr(module, func_name)(args, parser)
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/cli/main_init.py", line 155, in execute
        return initialize(
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/core/initialize.py", line 137, in initialize
        run_plan(plan2)
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/core/initialize.py", line 798, in run_plan
        result = globals()[step["function"]](
      File "/home/stud/rschwa/miniforge/lib/python3.10/site-packages/conda/core/initialize.py", line 1454, in init_sh_user
        rc_content = fh.read()
      File "/home/stud/rschwa/miniforge/lib/python3.10/codecs.py", line 322, in decode
        (result, consumed) = self._buffer_decode(data, self.errors, final)
    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe4 in position 313: invalid continuation byte

`$ /home/stud/rschwa/miniforge/bin/conda init --all`

  environment variables:
                 CIO_TEST=<not set>
               CONDA_ROOT=/home/stud/rschwa/miniforge
           CURL_CA_BUNDLE=<not set>
                 INFOPATH=/opt/software/gcc/gcc-11.2.0/share/info:
          LD_LIBRARY_PATH=/opt/software/gcc/gcc-11.2.0/lib:/opt/software/gcc/gcc-11.2.0/lib64:
               LD_PRELOAD=<not set>
                  MANPATH=/opt/software/gcc/gcc-11.2.0/share/man:
                NODE_PATH=/usr/lib/nodejs:/usr/lib/node_modules:/usr/share/javascript
                     PATH=/home/stud/rschwa/miniforge/bin:/home/stud/rschwa/miniforge/bin:/opt/s
                          oftware/gcc/gcc-
                          11.2.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/a
                          fs/tu-chemnitz.de/home/urz/r/rschwa/:/afs/tu-
                          chemnitz.de/home/urz/r/rschwa/bin:/afs/tu-
                          chemnitz.de/home/urz/r/rschwa/:/afs/tu-
                          chemnitz.de/home/urz/r/rschwa/bin
       REQUESTS_CA_BUNDLE=<not set>
            SSL_CERT_FILE=<not set>

     active environment : None
       user config file : /afs/tu-chemnitz.de/home/urz/r/rschwa/.condarc
 populated config files : /home/stud/rschwa/miniforge/.condarc
          conda version : 24.1.2
    conda-build version : not installed
         python version : 3.10.14.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=haswell
                          __conda=24.1.2=0
                          __glibc=2.24=0
                          __linux=4.9.0=0
                          __unix=0=0
       base environment : /home/stud/rschwa/miniforge  (writable)
      conda av data dir : /home/stud/rschwa/miniforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/stud/rschwa/miniforge/pkgs
                          /afs/tu-chemnitz.de/home/urz/r/rschwa/.conda/pkgs
       envs directories : /home/stud/rschwa/miniforge/envs
                          /afs/tu-chemnitz.de/home/urz/r/rschwa/.conda/envs
               platform : linux-64
             user-agent : conda/24.1.2 requests/2.31.0 CPython/3.10.14 Linux/4.9.0-8-amd64 debian/9.13 glibc/2.24 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.7
                UID:GID : 91994:4000
             netrc file : None
           offline mode : False


An unexpected error has occurred. Conda has prepared the above report.
If you suspect this error is being caused by a malfunctioning plugin,
consider using the --no-plugins option to turn off plugins.

Example: conda --no-plugins install <package>

Alternatively, you can set the CONDA_NO_PLUGINS environment variable on
the command line to run the command without plugins enabled.

Example: CONDA_NO_PLUGINS=true conda install <package>

If submitted, this report will be used by core maintainers to improve
future releases of conda.
Would you like conda to send this report to the core maintainers? [y/N]: y
Upload successful.

Thank you for helping to improve conda.
Opt-in to always sending reports (and not see this message again)
by running

    $ conda config --set report_errors true

rschwa@leander:/home/stud/rschwa$ echo $0
bash
```

Up to this point everything looked normal. The last line is because `echo $0` is recommended when conda init fails. 

The root of the issues might be the really old OS (Debian 9), something I cannot change right now. However, I could not find any info, that it is not supported anymore. Moreover, the error does not really look like something associated with too old OS. Anyway, I have no idea how to tackle it and would appreciate any help.

**Comment by milljm** [comment]
Almost looks like you might already have a Conda installation elsewhere?

Starting a new terminal with out having `PATH=/home/stud/rschwa/miniforge:$PATH` occurring, what happens when you invoke some `conda` commands? Namely, the following would be of interest:
```bash
conda info --all
```
If this returns `command not found`, well then I am wrong, and we need to focus elsewhere.

**Reply by WulfHans** [reply]
```
rschwa@leander:/home/stud/rschwa$ echo $PATH
/opt/software/gcc/gcc-11.2.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/afs/tu-chemnitz.de/home/urz/r/rschwa/:/afs/tu-chemnitz.de/home/urz/r/rschwa/bin:/afs/tu-chemnitz.de/home/urz/r/rschwa/:/afs/tu-chemnitz.de/home/urz/r/rschwa/bin
rschwa@leander:/home/stud/rschwa$ which conda
rschwa@leander:/home/stud/rschwa$ conda
bash: conda: Kommando nicht gefunden.
```

Last line is command not found in german ;)

**Reply by milljm** [reply]
aight, what if we ignore trying to initialize conda, and just use it... does that work?

```bash
export PATH=/home/stud/rschwa/miniforge/bin:$PATH
source activate /home/stud/rschwa/miniforge
conda info --all
conda list
```

Edit: sorry for the delayed response

p.s
>bash: conda: Kommando nicht gefunden.

oh my! i didn't even notice!!! lol I am so used to that sort of syntax output for commands not found I blew right past that it was in a different language.

**Reply by WulfHans** [reply]
I switched my attempts to another computer with Debian 12, because I wanted to avoid any problems coming from an ancient OS. That is why the path has changed slightly (stud_syrinx instead of stud). 
Unfortunately, the error is still the same. In other words: The old OS was NOT the problem. I am starting to suspect that the current conda version is just corrupt.

Anyway lets continue with the proposed tests:

```
rschwa@syrinx:/home/stud_syrinx/rschwa/moose_projects/install$ source activate /home/stud_syrinx/rschwa/miniforge/
(/home/stud_syrinx/rschwa/miniforge) rschwa@syrinx:/home/stud_syrinx/rschwa/moose_projects/install$ conda info --all

     active environment : base
    active env location : /home/stud_syrinx/rschwa/miniforge
            shell level : 2
       user config file : /afs/tu-chemnitz.de/home/urz/r/rschwa/.condarc
 populated config files : /home/stud_syrinx/rschwa/miniforge/.condarc
          conda version : 24.1.2
    conda-build version : not installed
         python version : 3.10.14.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=haswell
                          __conda=24.1.2=0
                          __glibc=2.36=0
                          __linux=6.1.0=0
                          __unix=0=0
       base environment : /home/stud_syrinx/rschwa/miniforge  (writable)
      conda av data dir : /home/stud_syrinx/rschwa/miniforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/stud_syrinx/rschwa/miniforge/pkgs
                          /afs/tu-chemnitz.de/home/urz/r/rschwa/.conda/pkgs
       envs directories : /home/stud_syrinx/rschwa/miniforge/envs
                          /afs/tu-chemnitz.de/home/urz/r/rschwa/.conda/envs
               platform : linux-64
             user-agent : conda/24.1.2 requests/2.31.0 CPython/3.10.14 Linux/6.1.0-18-amd64 debian/12.5 glibc/2.36 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.7
                UID:GID : 91994:4000
             netrc file : None
           offline mode : False


# conda environments:
#
base                  *  /home/stud_syrinx/rschwa/miniforge

sys.version: 3.10.14 | packaged by conda-forge | (mai...
sys.prefix: /home/stud_syrinx/rschwa/miniforge
sys.executable: /home/stud_syrinx/rschwa/miniforge/bin/python
conda location: /home/stud_syrinx/rschwa/miniforge/lib/python3.10/site-packages/conda
conda-build: None
conda-env: /home/stud_syrinx/rschwa/miniforge/bin/conda-env
user site dirs: 

CIO_TEST: <not set>
CONDA_DEFAULT_ENV: /home/stud_syrinx/rschwa/miniforge
CONDA_EXE: /home/stud_syrinx/rschwa/miniforge/bin/conda
CONDA_PREFIX: /home/stud_syrinx/rschwa/miniforge
CONDA_PREFIX_1: /home/stud_syrinx/rschwa/miniforge
CONDA_PROMPT_MODIFIER: (/home/stud_syrinx/rschwa/miniforge) 
CONDA_PYTHON_EXE: /home/stud_syrinx/rschwa/miniforge/bin/python
CONDA_ROOT: /home/stud_syrinx/rschwa/miniforge
CONDA_SHLVL: 2
CURL_CA_BUNDLE: <not set>
LD_PRELOAD: <not set>
PATH: /home/stud_syrinx/rschwa/miniforge/bin:/home/stud_syrinx/rschwa/miniforge/condabin:/home/stud_syrinx/rschwa/miniforge/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/afs/tu-chemnitz.de/home/urz/r/rschwa:/afs/tu-chemnitz.de/home/urz/r/rschwa/bin
REQUESTS_CA_BUNDLE: <not set>
SSL_CERT_FILE: <not set>
```
Looks as if its actually working. The environment is listed as (/home/stud_syrinx/rschwa/miniforge) instead of just (base) in the command prompt, but the rest looks normal to me. 

Lets see the contents of the environment:
```
(/home/stud_syrinx/rschwa/miniforge) rschwa@syrinx:/home/stud_syrinx/rschwa/moose_projects/install$ conda list
# packages in environment at /home/stud_syrinx/rschwa/miniforge:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
archspec                  0.2.3              pyhd8ed1ab_0    conda-forge
boltons                   23.1.1             pyhd8ed1ab_0    conda-forge
brotli-python             1.1.0           py310hc6cd4ac_1    conda-forge
bzip2                     1.0.8                hd590300_5    conda-forge
c-ares                    1.27.0               hd590300_0    conda-forge
ca-certificates           2024.2.2             hbcca054_0    conda-forge
certifi                   2024.2.2           pyhd8ed1ab_0    conda-forge
cffi                      1.16.0          py310h2fee648_0    conda-forge
charset-normalizer        3.3.2              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
conda                     24.1.2          py310hff52083_0    conda-forge
conda-libmamba-solver     24.1.0             pyhd8ed1ab_0    conda-forge
conda-package-handling    2.2.0              pyh38be061_0    conda-forge
conda-package-streaming   0.9.0              pyhd8ed1ab_0    conda-forge
distro                    1.9.0              pyhd8ed1ab_0    conda-forge
fmt                       10.2.1               h00ab1b0_0    conda-forge
icu                       73.2                 h59595ed_0    conda-forge
idna                      3.6                pyhd8ed1ab_0    conda-forge
jsonpatch                 1.33               pyhd8ed1ab_0    conda-forge
jsonpointer               2.4             py310hff52083_3    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
krb5                      1.21.2               h659d440_0    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libarchive                3.7.2                h2aa1ff5_1    conda-forge
libcurl                   8.6.0                hca28451_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 hd590300_2    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 13.2.0               h807b86a_5    conda-forge
libgomp                   13.2.0               h807b86a_5    conda-forge
libiconv                  1.17                 hd590300_2    conda-forge
libmamba                  1.5.7                had39da4_0    conda-forge
libmambapy                1.5.7           py310h39ff949_0    conda-forge
libnghttp2                1.58.0               h47da74e_1    conda-forge
libnsl                    2.0.1                hd590300_0    conda-forge
libsolv                   0.7.28               hfc55251_0    conda-forge
libsqlite                 3.45.2               h2797004_0    conda-forge
libssh2                   1.11.0               h0841786_0    conda-forge
libstdcxx-ng              13.2.0               h7e041cc_5    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libxcrypt                 4.4.36               hd590300_1    conda-forge
libxml2                   2.12.6               h232c23b_0    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
lz4-c                     1.9.4                hcb278e6_0    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
mamba                     1.5.7           py310h51d5547_0    conda-forge
menuinst                  2.0.2           py310hff52083_0    conda-forge
ncurses                   6.4.20240210         h59595ed_0    conda-forge
openssl                   3.2.1                hd590300_1    conda-forge
packaging                 24.0               pyhd8ed1ab_0    conda-forge
pip                       24.0               pyhd8ed1ab_0    conda-forge
platformdirs              4.2.0              pyhd8ed1ab_0    conda-forge
pluggy                    1.4.0              pyhd8ed1ab_0    conda-forge
pybind11-abi              4                    hd8ed1ab_3    conda-forge
pycosat                   0.6.6           py310h2372a71_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
python                    3.10.14         hd12c33a_0_cpython    conda-forge
python_abi                3.10                    4_cp310    conda-forge
readline                  8.2                  h8228510_1    conda-forge
reproc                    14.2.4.post0         hd590300_1    conda-forge
reproc-cpp                14.2.4.post0         h59595ed_1    conda-forge
requests                  2.31.0             pyhd8ed1ab_0    conda-forge
ruamel.yaml               0.18.6          py310h2372a71_0    conda-forge
ruamel.yaml.clib          0.2.8           py310h2372a71_0    conda-forge
setuptools                69.2.0             pyhd8ed1ab_0    conda-forge
tk                        8.6.13          noxft_h4845f30_101    conda-forge
tqdm                      4.66.2             pyhd8ed1ab_0    conda-forge
truststore                0.8.0              pyhd8ed1ab_0    conda-forge
tzdata                    2024a                h0c530f3_0    conda-forge
urllib3                   2.2.1              pyhd8ed1ab_0    conda-forge
wheel                     0.43.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
yaml-cpp                  0.8.0                h59595ed_0    conda-forge
zstandard                 0.22.0          py310h1275a96_0    conda-forge
zstd                      1.5.5                hfc55251_0    conda-forge
```
No idea how it should look like, does not look completely wrong.
I could continue without initializing now - is that viable or a really bad idea? What does conda init even do apart from putting a block of code into .bashrc?

**Reply by GiudGiud** [reply]
Should be the moose environment in the prompt. 

You could just delete conda and re install. It takes a lot less time than trying to fix it. Make sure to use the miniforge one we point to in our website

**Reply by GiudGiud** [reply]
Should be the moose environment in the prompt. 

You could just delete conda and re install. It takes a lot less time than trying to fix it. Make sure to use the miniforge one we point to in our documentation

**Reply by GiudGiud** [reply]
It should show as (moose) in the console prompt if you installed conda and ran conda init then activated the moose environment. 

And the list should contain a few moose- prefixed packages that are the dependencies of moose

i don’t know all that conda init does. You can probably find out in the conda documentation

**Reply by WulfHans** [reply]
I used exactly the command from the moose install page. I also deleted and reinstalled it several times and tried  two different PCs. The error is persistent. 
The different environment (instead of (moose) ) is due to the fact, that I did not continue with installation after `conda init` failed.
However, being curious and impatient, I did exactly that in the meantime: Ignore the conda init completely and just use it. And it seems to work fine. It's a bit scary, because this is clearly not the intended way - but I am not really motivated to experiment now, because I fear messing up a working state.

**Reply by milljm** [reply]
Sorry for the delayed response, it was our three day weekend.

>```bash
>export PATH=/home/stud/rschwa/miniforge/bin:$PATH
>source activate /home/stud/rschwa/miniforge
>conda info --all
>conda list
>```
You did exactly what I asked, and your print out is exactly what I expected (there shouldn't be any `(moose)` prompt). And it worked. So at least we have a work-around in the event we can't figure out `conda init`.

Next, lets do create the moose environment and see if we can activate that:
```bash
export PATH=/home/stud_syrinx/rschwa/miniforge/bin:$PATH
source activate /home/stud_syrinx/rschwa/miniforge
conda config --add channels https://conda.software.inl.gov/public
conda create -n moose moose-dev
source activate /home/stud_syrinx/rschwa/miniforge/envs/moose
```
If Conda successfully activate the moose environment, you should be able to build/run MOOSE. But I think the issue still remains "why doesn't `conda init bash` work". Does it now? After everything we've done?

```bash
conda init bash
```

**Reply by WulfHans** [reply]
```
rschwa@syrinx:~$ echo $PATH
/home/stud_syrinx/rschwa/miniforge/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/afs/tu-chemnitz.de/home/urz/r/rschwa/:/afs/tu-chemnitz.de/home/urz/r/rschwa/bin
rschwa@syrinx:~$ source activate /home/stud_syrinx/rschwa/miniforge/envs/moose
(moose) rschwa@syrinx:~$
```

Works fine, and compiling and using moose does work as well. So this works as workaround. 

However, `conda init bash` still fails with exactly the same error as shown above (I will save the space for quoting it again).
Also, just `conda init moose` fails with the message `Please call conda init first` - but the command you gave me above using source activate works.

**Reply by milljm** [reply]
For your peace of mind, all of our build machines running the show here: https://civet.inl.gov, are using the same methods to activate the moose environment as you are at this point. Its a very clean way to do it. that is: it won't pick up anything stray from your bash/zsh/tcsh profiles by accident (and that is why we do it in our build environments).

Anyways! Not being able to run `conda activate` as a regular user is quite the inconvenience! Can we look at your bash (or whatever shell you are using) profiles? I think that would be the next logical step.

To prove if it is actually an environment profile issue getting in the way, does this work:

```bash
env -i bash --norc --noprofile
# some needed stuff
export HOME=/home/stud_syrinx
export PATH=/bin:/usr/bin:/sbin
export TERM=xterm-256color

# and now the troubleshooting:
export PATH=/home/stud_syrinx/rschwa/miniforge/bin:$PATH
conda init bash
```

Something I want to point out:
>Also, just conda init moose 

I am not sure this is a valid command to be honest. `conda init <insert your terminal shell type here>` is what should follow the `init` argument. Be it, ZSH, TCSH, BASH, etc type shells.

**Reply by WulfHans** [reply, marked as answer]
I solved it!

The bashrc template supplied by our central administration had some non-UTF8  symbols in a **comment** . German language has this wonderful thing called umlaut and some brilliant mind put them in the disclaimer and instruction comment block at the top (which I have not read carefully for years and have not bothered to copy here). It looks fine in any editor,  but conda does not just append its block and is happy, it just dies to that. 

Bottom line: 
1. It was our own fault, but conda could be more error-tolerant or clear in its error messages.
2. I absolutely despise any encoding that is not utf-8
3. Thanks for your patience and support once more! I'll mark it as fixed.

**Reply by milljm** [reply]
We were certainly heading in the right direction! Cheers!
